File: src/COS_Map.pl
#!/usr/bin/perl -w
use strict;

use cosmic;

sub ShowMapLesson {
  my $txt = "";
  
  $txt .= "# MATH show map function for applying a function across the elements of a list\n";

  $txt .= ShowLine(Op2("define",
		       "map",
		       ProcMultiple(["p", "lst"],
				    Op("if",
				       Op2(">",
					   Op1("list-length", Ref("lst")),
					   0),
				       Op2("prepend",
					   Op("p",Op1("head",Ref("lst"))),
					   Op("map",
					      Ref("p"),
					      Op1("tail",Ref("lst")))),
				       ShowList()))));

  for (my $i=0; $i<4; $i++)
    {
      my @list = prand(20,$i+3);
      my @out = ();
      foreach my $n (@list)
	{
	  push(@out,2*$n);
	}
      $txt .= ShowLine(Op2("list=",
			   Op("map",
			      Proc("x",
				   Op2("*",Ref("x"),2)),
			      ShowList(@list)),
			   ShowList(@out)));
    }
     
  for (my $i=0; $i<4; $i++)
    {
      my @list = prand(20,$i+3);
      my @out = ();
      foreach my $n (@list)
	{
	  push(@out,42);
	}
      $txt .= ShowLine(Op2("list=",
			   Op("map",
			      Proc("x",
				   42),
			      ShowList(@list)),
			   ShowList(@out)));
    }

  $txt .= ShowLine(Op2("define",
		       "crunch",
		       ProcMultiple(["p", "lst"],
				    Op("if",
				       Op2(">=",
					   Op1("list-length", Ref("lst")),
					   2),
				       Apply("p",
					     Op1("head",Ref("lst")),
					     Op("crunch",
						Ref("p"),
						Op1("tail",Ref("lst")))),
				       Op("if",
					  Op2("=",
					      Op1("list-length", Ref("lst")),
					      1),
					  Op1("head",Ref("lst")),
					  Ref("undefined"))))));
  for (my $i=0; $i<4; $i++)
    {
      my @list = prand(20,$i+3);
      my @out = ();
      my $sum = 0;
      foreach my $n (@list)
	{
	  $sum += $n;
	}
      
      $txt .= ShowLine(Op2("=",
			   Op("crunch",
			      Ref("+"),
			      ShowList(@list)),
			   $sum));
    }
  

  return $txt;
};


ShowLesson(ShowMapLesson());

File: msg/COS_Map.js
#!/usr/bin/env node

const cos = require("./cosmic");

cos.language(2);
cos.seed(42);

cos.intro("map");
cos.add(`
define map | ? x:? | ? x:list |
  if (= 0 | list-length $x:list) (list 0) |
  prepend (x:? | head $x:list) (map $x:? | tail $x:list)`);

for (let i = 0; i < 4; i++) {
  const lst = cos.prand(20, i + 3);
  const out = lst.map(v => 2 * v);
  cos.add(`list= (${cos.listExpression(out)}) | map (? x | * $x 2) | ${cos.listExpression(lst)}`);
}

for (let i = 0; i < 4; i++) {
  const lst = cos.prand(20, i + 3);
  const out = lst.map(v => 42);
  cos.add(`list= (${cos.listExpression(out)}) | map (? x 42) | ${cos.listExpression(lst)}`);
}

cos.intro("reduce");
cos.add(`
define reduce | ? x:? | ? x:list |
  if (= 0 | list-length $x:list) $undefined |
  if (= 1 | list-length $x:list) (head $x:list) |
  x:? (head $x:list) (reduce $x:? | tail $x:list)`);

for (let i = 0; i < 4; i++) {
  const lst = cos.prand(20, i + 3);
  const out = lst.reduce((a, b) => a + b);
  cos.add(`= ${out} | reduce $+ | ${cos.listExpression(lst)}`);
}
